### DISCUSSION TITLE
REST API Authorization

### DISCUSSION THREAD

**Discussion by queza85** [exoplatform/exo-community#50] [Q&A]
Hi 

I am trying to use social api like below capture.
<img width="526" alt="image" src="https://github.com/exoplatform/exo-community/assets/2861131/87a3b6c3-39ca-4cfe-bc16-df6e2d414034">

So I found the way how to use authorization based on old document(new document has no information).
I tried same way to generate authorization value 

<img width="661" alt="image" src="https://github.com/exoplatform/exo-community/assets/2861131/e7b375b6-0387-40e7-829a-bca9e0f0a7d6">
<way to use rest api>

<img width="343" alt="image" src="https://github.com/exoplatform/exo-community/assets/2861131/3f619871-3d62-4def-af1d-172b75ce7018">
<generate authorization key>

But I have issue to autholization.

<img width="660" alt="image" src="https://github.com/exoplatform/exo-community/assets/2861131/d3c02b95-a810-4aba-ba3c-cf57d7e49cc4">

<img width="803" alt="image" src="https://github.com/exoplatform/exo-community/assets/2861131/9d4e3555-f2d6-4a02-8839-f21f3686cb86">
<using bytes from plain text and then encode base64>

Can I get any advise to call this api?

**Comment by queza85** [comment]
Sorry one more question that I add user via 'csv file' in user management screen.
it shows like this error too.

<img width="1564" alt="image" src="https://github.com/exoplatform/exo-community/assets/2861131/9d81dd28-e567-438d-8933-8559e4ac526f">

the sample is from your example. but what is the meaning key value? 
<img width="607" alt="image" src="https://github.com/exoplatform/exo-community/assets/2861131/7ad267af-a99f-4867-a8fd-3d76e074f233">

**Reply by queza85** [reply]
Hi @rdenarie 
Can I get any help or document for this?

**Reply by rdenarie** [reply]
Hello,

About API access, the url to use basic authentication is /rest/private/xxx instead of /portal/rest/xxx
So in your case, you should access to 
/rest/private/v1/social/users

In postman, in authentication tab, you can choose authentication method, for example BASIC, and then fill username and password. Postman will encode it directly, no need to encode it manually : 
![image](https://github.com/exoplatform/exo-community/assets/807839/5aaff6a5-5c6a-4a4d-b551-a97991366f92)



About import by CSV : 
The csv format must be

userName,firstName,lastName,password,email
john,John,Smith,Password1234,jsmith@acme.com

- userName instead of login
- key must be removed
- check that separator is ',' and not ';'

**Reply by queza85** [reply]
@rdenarie  thanks for your inform.

If we develop java base, is this still working from your old document?

<img width="641" alt="image" src="https://github.com/exoplatform/exo-community/assets/2861131/5fa6d744-8976-4d68-a66f-64073dbbad2b">

In my case, we connected with Keycloak. so we use our own users. but the root is not existed.
So can we use another certain user like capture? or have to use only root account?

<img width="1634" alt="image" src="https://github.com/exoplatform/exo-community/assets/2861131/3cdbc1fb-a906-4bfb-9390-1fbc59d80470">

**Reply by rdenarie** [reply]
In fact, root user still exists, but only on eXo side. He is not in your keycloack user database, but as he exists in eXo, he can connect.

The basic method still work, but only when using url like /rest/private/... instead of /portal/rest/...

To use API with a keycloack user, you will not be able to use basic authentication. You will have to simulate the complete login phase :
Access to the same url as when the user click on the keycloak button on login page
Make the authentication
Follow the redirect, and get the JSESSION_ID cookie
Then use the cookie to make the API call

**Reply by queza85** [reply]
Do you mean, I can use root user even I turn on SSO?
Actually, I'm okay any user if we can use the Rest API.

So, when I try this to use our root account which can be login without SSO
It is shown like this error from exo server.

![image](https://github.com/exoplatform/exo-community/assets/2861131/c1fef3ae-e955-438f-a85a-f7fb26fe5b56)
<Call 'rest/private/v1/social/users', it's working without SSO config>

![image](https://github.com/exoplatform/exo-community/assets/2861131/b1f455e6-3dd8-4474-b1db-e6d8d20acd33)
<Response: Unauthorized error>

<img width="898" alt="image" src="https://github.com/exoplatform/exo-community/assets/2861131/193a9360-fd33-48a9-83f3-ef37a96f75db">
<Server Log for error>
=> Does it need to install something? before I setup SSO via SAML, I added package based on document.

**Comment by rdenarie** [comment]
I'm sorry, I mix different contexts.
I though you use OpenId. 
In this case, user only in eXo like root can connect.

With SAML, user must exists in the Identity Provider, so root cannot connect.

**Reply by queza85** [reply]
any user is okay? or it has some permission to access RESTAPI?
I just tried to use certain account has MEMBER role.
But same like above issue.

**Reply by rdenarie** [reply]
It depends of API.
For example, the one to get users list is accessible to all users. 

In fact, you speak about role MEMBER. Can you explain what do you mean for this role ? 

When you log in eXo, an go in users and groups administration, you can groups : 
- /platform/administrators
- /platform/users
- ...

For each group, users are associated with a role : 
- manager 
- member
- ...

So a permission is an association between a role and a group : 
- member:/platform/users
- manager:/platform/administrators

To access to /v1/social/users, the user must be in the group /platform/users, neverming his role.
Can you ensure that, when logging in the platform, the user you use is present in this group ?


> I just tried to use certain account has MEMBER role.

Can you explain what you tried ?

**Reply by queza85** [reply]
I would like to retrieve user list and save user also.
So I would like to use GET /v1/social/users and POST /v1/social/user.

I tested based on your guide like capture.
I added test account to group for admin and user. 
That is home screen after login keycloak.

![image](https://github.com/exoplatform/exo-community/assets/2861131/0b31dc0b-79e7-44ac-9bfc-d8d62f55a93e)
==>successfully keycloak login as a admin group

<img width="1136" alt="image" src="https://github.com/exoplatform/exo-community/assets/2861131/768f36a2-997d-4059-aac0-35a582cac353">
==>fail to call api even use certain admin group account

![image](https://github.com/exoplatform/exo-community/assets/2861131/9540802c-77d5-420d-843f-a94f07e47d3f)
==>exo log when I call api, NoClassDefFoundError


Please, advise for this. thanks

**Comment by rdenarie** [comment]
In fact, the BASIC authentication will not work with SAML authentication. 

You will need to make 2 requests (at least) : one for the login, and get the result JSESSIONID cookie, and the second for the api call, using the cookie.

**Reply by queza85** [reply]
Ah.. I see. 
So if I follow previous your guide for login, I can use only click by user?
I would like to make it automatically call the rest api. but it looks impossible if user has to click login page directly...

"Access to the same url as when the user click on the keycloak button on login page"
Make the authentication
Follow the redirect, and get the JSESSION_ID cookie
Then use the cookie to make the API call"

**Reply by rdenarie** [reply]
It is not mandatory to make click by a user.
I dont know exactly which tool you use to make API call, but your tool can make different requests : 
- first request is the login request. In the response you read the cookie
- second request use the cookie, then call the API you want.

**Reply by queza85** [reply]
I developed spring boot application.
If I use this, how can I make first request in back-end side? I understood this is done on UI side by the user, doesn't it?

**Reply by rdenarie** [reply]
When you make a login in your browser, and observ with developer tool the network tab, you will see the login request, which is a POST. So you can reproduce this request with your spring boot app.

**Reply by queza85** [reply]
ah. I see.
I will try to catch when I make a login in the keycloak.

**Reply by paljitender** [reply]
@rdenarie,

Getting 401 unauthorize while fetching the spaces data based on the ID

using Following info:

URL :- http://localhost:8080/rest/private/v1/social/spaces/gdp

Adding the screen shot below;


![image](https://github.com/exoplatform/exo-community/assets/113160963/1f4b8bc0-4020-4da0-80c1-c410754e4900)

**Reply by rdenarie** [reply]
What is 'gdp' ? is it a space Name ? 

Do request /rest/private/v1/social/spaces is working ?

**Comment by Chiranjeevi-d** [comment]
If the user login through openId then how he can access the exo platfrom api's. Is there any other way of login is supported other than Basic authentication.

**Reply by rdenarie** [reply]
Hi,
To call directly eXo API, with only one request, there is only BASIC auth currently.

If you want to use openId, you will need to reproduce the sequence of requests done in a browser, by storing and reusing cookies between requests.

**Comment by Chiranjeevi-d** [comment]
Could you plz share any supportive document how to this. Actually we have an application authentication is done through ad  with the same credentials need login into exo platform is it possible.



Chiranjeevi D




---- On Mon, 19 Aug 2024 13:24:22 +0530 ***@***.*** wrote ----



Hi,
To call directly eXo API, with only one request, there is only BASIC auth currently.

If you want to use openId, you will need to reproduce the sequence of requests done in a browser, by storing and reusing cookies between requests.

**Comment by Chiranjeevi-d** [comment]
Could you plz share any supportive document how to do this. 
      Actually we have an application, authentication is done through  microsoft ad  with the same credentials need login into exo platform is it possible to do this 



Chiranjeevi D




---- On Mon, 19 Aug 2024 13:24:22 +0530 ***@***.*** wrote ----



Hi,
To call directly eXo API, with only one request, there is only BASIC auth currently.

If you want to use openId, you will need to reproduce the sequence of requests done in a browser, by storing and reusing cookies between requests.

**Comment by rdenarie** [comment]
I have not documentation about that.
It is an integration subject. 

You need to analyze requests done during a login process in a browser (use browser dev tool). 
Then reproduce theses requests in a programatical manner to replay the login processus.

**Comment by Chiranjeevi-d** [comment]
Thanks for your support !!!




Chiranjeevi D




---- On Mon, 19 Aug 2024 13:39:26 +0530 ***@***.*** wrote ----



I have not documentation about that.
It is an integration subject.

You need to analyze requests done during a login process in a browser (use browser dev tool).
Then reproduce theses requests in a programatical manner to replay the login processus.
